# Example 1: Linear Regression

Suppose you have the following dataset (taken from Module 2). 

| $X_1$  | $X_2$  | $X_3$  | $y$  |
|--------|--------|--------|--------|
| 1.2 | 2.3 | 3.1 | 8.2 |
| 2.0 | 1.8 | 2.6 | 7.5 |
| 3.1 | 2.7 | 4.0 | 10.3 |
| 4.0 | 3.9 | 5.1 | 13.2 |
| 5.2 | 4.7 | 6.0 | 15.1 |
| 6.0 | 5.9 | 7.2 | 17.2 |
| 7.1 | 6.8 | 8.1 | 19.0 |

We define 

$$\mathbf{X} =
\begin{bmatrix}
1 & 1.2 & 2.3 & 3.1 \\
1 & 2.0 & 1.8 & 2.6 \\
1 & 3.1 & 2.7 & 4.0 \\
1 & 4.0 & 3.9 & 5.1 \\
1 & 5.2 & 4.7 & 6.0 \\
1 & 6.0 & 5.9 & 7.2 \\
1 & 7.1 & 6.8 & 8.1 \\
\end{bmatrix} \quad \text{ and } \quad
\mathbf{y} = \begin{bmatrix}
8.2 \\
7.5 \\
10.3 \\
13.2 \\
15.1 \\
17.2 \\
19.0 \\
\end{bmatrix}$$

Then 

$$\mathbf{X}^T =
\begin{bmatrix}
1 & 1 & 1 & 1 & 1 & 1 & 1\\
1.2 & 2.0 & 3.1 & 4.0 & 5.2 & 6.0 & 7.1 \\
2.3 & 1.8 & 2.7 & 3.9 & 4.7 & 5.9 & 6.8 \\
3.1 & 2.6 & 4.0 & 5.1 & 6.0 & 7.2 & 8.1 \\
\end{bmatrix}, \quad \quad 
\mathbf{X}^T \mathbf{X} =
\begin{bmatrix}
7 & 28.6 & 28.1 & 36.1 \\
28.6 & 144.5 & 138.45 & 173.63 \\
28.1 & 138.45 & 134.17 & 168.26 \\
36.1 & 173.63 & 168.26 & 211.83 \\
\end{bmatrix}, \quad \text{and}$$ $$ \mathbf{X}^T \mathbf{y} =
\begin{bmatrix}
90.5 \\
426.19 \\
413.3 \\
521.78 \\
\end{bmatrix}
$$

The inverse of $(\mathbf{X}^T \mathbf{X})^{-1}$ is:

$$(\mathbf{X}^T \mathbf{X})^{-1} =
\begin{bmatrix}
8.8541859 & 2.16305293 & 7.59490793 & -9.31466836 \\
2.16305293 & 1.14062489 & 1.22606673 & -2.2774437 \\
7.59490793 & 1.22606673 & 9.09088261 & -9.52032314 \\
-9.31466836 & -2.2774437 & -9.52032314 & 11.0210152 \\
\end{bmatrix}.
$$

We now have all the elements to find $\mathbf{\hat{\beta}}$:

\begin{align*} 
\mathbf{\hat{\beta}} &= (\mathbf{X}^T \mathbf{X})^{-1}\mathbf{X}^T \mathbf{y} \\
&= \begin{bmatrix}
8.8541859 & 2.16305293 & 7.59490793 & -9.31466836 \\
2.16305293 & 1.14062489 & 1.22606673 & -2.2774437 \\
7.59490793 & 1.22606673 & 9.09088261 & -9.52032314 \\
-9.31466836 & -2.2774437 & -9.52032314 & 11.0210152 \\
\end{bmatrix}
\begin{bmatrix}
90.5 \\
426.19 \\
413.3 \\
521.78 \\
\end{bmatrix} \\
&= \begin{bmatrix}
1.94314074 \\
0.28801913 \\
-0.37587751 \\
2.19453811 \\
\end{bmatrix}
\end{align*}

This means our model is of the form 

$$ \hat{Y} = 1.94314074 + 0.28801913 X_1 + -0.37587751 X_2 + 2.19453811  X_3 $$

Pay careful attention to the order of the values of $\beta$ and the order of the columns in $\mathbf{x}$; these should correspond with one another!

## Explicit Computation in Code 

We can compute this by hand, with a graphing calculator or explicitly by code. Below is the code for this explicit computation in Python. 

In [1]:
# numpy is a library for arrays in python
# and it is often abbreviated to "np"
import numpy as np

This is the dataset from the problem. We encode it into a numpy array which tends to be more efficient than a list.

In [2]:
# Create matrix x
X = np.array([[1, 1.2, 2.3, 3.1],
              [1, 2.0, 1.8, 2.6],
              [1, 3.1, 2.7, 4.0],
              [1, 4.0, 3.9, 5.1],
              [1, 5.2, 4.7, 6.0],
              [1, 6.0, 5.9, 7.2],
              [1, 7.1, 6.8, 8.1]])

# Create matrix y
y = np.array([[8.2],
              [7.5],
              [10.3],
              [13.2],
              [15.1],
              [17.2],
              [19.0]])

In [3]:
# Calculate x^T
X_t = np.transpose(X)

# Calculate the product of x^T and x
prod_X_t_X = np.dot(X_t, X)

# Calculate the inverse of the product of X^T and X
inv_prod_X_t_X = np.linalg.inv(prod_X_t_X)

# Calculate the product of X^T and y
prod_X_t_y = np.dot(X_t, y)

# Calculate coefficients
beta = np.dot(inv_prod_X_t_X, prod_X_t_y)

In [4]:
# Print the results
print("X^T:")
print(X_t)
print("\nProduct of X^T and X:")
print(prod_X_t_X)
print("\nInverse of the product of X^T and X, (X^T X)^{-1}:")
print(inv_prod_X_t_X)
print("\nProduct of X^T and y:")
print(prod_X_t_y)
print("\nCoefficients for Linear Regression:")
print(beta)

X^T:
[[1.  1.  1.  1.  1.  1.  1. ]
 [1.2 2.  3.1 4.  5.2 6.  7.1]
 [2.3 1.8 2.7 3.9 4.7 5.9 6.8]
 [3.1 2.6 4.  5.1 6.  7.2 8.1]]

Product of X^T and X:
[[  7.    28.6   28.1   36.1 ]
 [ 28.6  144.5  138.45 173.63]
 [ 28.1  138.45 134.17 168.26]
 [ 36.1  173.63 168.26 211.83]]

Inverse of the product of X^T and X, (X^T X)^{-1}:
[[ 8.8541859   2.16305293  7.59490793 -9.31466836]
 [ 2.16305293  1.14062489  1.22606673 -2.2774437 ]
 [ 7.59490793  1.22606673  9.09088261 -9.52032314]
 [-9.31466836 -2.2774437  -9.52032314 11.0210152 ]]

Product of X^T and y:
[[ 90.5 ]
 [426.19]
 [413.3 ]
 [521.78]]

Coefficients for Linear Regression:
[[ 1.94314074]
 [ 0.28801913]
 [-0.37587751]
 [ 2.19453811]]


## Coding Linear Regression Quickly

There are several libraries that support the computation of linear regression. Below, we will demonstrate two that we will use often:
1. statsmodels
2. scikit-learn

Both packages are commonly used and have different strengths. Statsmodels is specifically designed from a statistics perspective. Scikit-Learn is a standardized model-building software that is ubiquitous in machine learning. 

### Statsmodels approach

In [5]:
import statsmodels.api as sm

In [6]:
# Fit a linear regression model
model = sm.OLS(y, X).fit()

# Note: we are using the various defined the previous section

In [7]:
print(model.params)

[ 1.94314074  0.28801913 -0.37587751  2.19453811]


Above are the coefficients of the model corresponding to the function 
$$\hat y = 1.94314074 + 0.28801913 X_1 - 0.37587751 X_2 + 2.19453811 X_3$$

## Scikit Learn approach

In [8]:
from sklearn.linear_model import LinearRegression

In [9]:
# Create a LinearRegression model
model = LinearRegression()

# Fit the model to the data
model.fit(X, y)

LinearRegression()

In [10]:
# Print the intercept and coefficients
print("Intercept:", model.intercept_)
print("Coefficients:", model.coef_)

Intercept: [1.94314074]
Coefficients: [[ 0.          0.28801913 -0.37587751  2.19453811]]


Scikit-learn separates out intercepts from coefficients and automatically includes the column we need to compute the intercept for the model. Because our $X$ is defined with this extra column of 1s, the corresponding coefficient is 0 and the intercept is separate. If we coded this more in-line with how scikit-learn is standardized, we would do this: 

In [11]:
# Create matrix x
X = np.array([[1.2, 2.3, 3.1],
              [2.0, 1.8, 2.6],
              [3.1, 2.7, 4.0],
              [4.0, 3.9, 5.1],
              [5.2, 4.7, 6.0],
              [6.0, 5.9, 7.2],
              [7.1, 6.8, 8.1]])

# Create matrix y
y = np.array([[8.2],
              [7.5],
              [10.3],
              [13.2],
              [15.1],
              [17.2],
              [19.0]])

# Create a LinearRegression model
model = LinearRegression()

# Fit the model to the data
model.fit(X, y)

# Print the intercept and coefficients
print("Intercept:", model.intercept_)
print("Coefficients:", model.coef_)

Intercept: [1.94314074]
Coefficients: [[ 0.28801913 -0.37587751  2.19453811]]


These coefficients and intercept corresponding to the function 
$$\hat y = 1.94314074 + 0.28801913 X_1 - 0.37587751 X_2 + 2.19453811 X_3$$

## Statistical Tools

When we wish to perform statistical tests, this turns out to be an easy exercise with the statsmodels package. 

In [12]:
# Standard libraries
import numpy as np
import statsmodels.api as sm

For this example, we will create a dataset using random number generation. Our true model is $$y = 2 + 0.3X.$$

In [13]:
# Create a synthetic dataset
np.random.seed(1)
X = 2.5 * np.random.randn(100) + 3      # Independent variable
epsilon = 0.5 * np.random.randn(100)    # Irreducible error
y = 2 + 0.3 * X + epsilon               # Dependent variable

Within the code below, we compute the linear regression model. Notice that since we are not manually constructing the matrix $X$, we need to specify that a coefficient must be added (this is mathematically equivalent to adding a column of 1s). 

In [14]:
#-------------- Compute Linear Regression Stats --------------# 

# Add a constant to the independent variable
X = sm.add_constant(X)

# Fit the regression model
model = sm.OLS(y, X).fit()

# Print out the statistics including t-values
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.698
Model:                            OLS   Adj. R-squared:                  0.695
Method:                 Least Squares   F-statistic:                     226.9
Date:                Sat, 25 Nov 2023   Prob (F-statistic):           3.00e-27
Time:                        17:00:49   Log-Likelihood:                -65.124
No. Observations:                 100   AIC:                             134.2
Df Residuals:                      98   BIC:                             139.5
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.0163      0.082     24.716      0.0

Based on the chart above, we see that the estimated model is $$\hat y = 2.0163 + 0.3191X.$$ The chart also provides information about whether the coefficient and intercept are statistically significant. We see that the p-value for both is estimated to be $0.000$, which means that it is so small, it is rounded down to 0 when considering 3 decimal places. The $t$-scores are quite large, so this is not surprising. 

The $F$ statistic is also computed with respect to the *entire* model. Here, we are considering both the coefficient and the intercept. We see that our $F$-score is also large, which corresponds to a very small p-value ($3.00 \times 10^{-27}$). 

We also get several other important metrics, including the coefficient of determination, $R^2$, which is estimated to be 0.698. This means that our model effectively captures 69.8% of the variation seen in $y$. 

Finally, it is important to highlight that with synthetic data, we know the true model and we can see that the estimated model, while close, is different. This is because all of our estimation practices will be effected by the noise expected in the data. The noisier the data, the larger the difference between the estimated model and the true model. This is also why smaller datasets are not ideal. The smaller the dataset, the more impacted it will be by the noise present. 